src/appleseed/renderer/kernel/lighting/backwardlightsampler.h

```diff
@@ -0,0 +1,260 @@
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2010-2013 Francois Beaune, Jupiter Jazz Limited
+// Copyright (c) 2014-2017 Francois Beaune, The appleseedhq Organization
+// Copyright (c) 2017 Petra Gospodnetic, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_RENDERER_KERNEL_LIGHTING_BACKWARDLIGHTSAMPLER_H
+#define APPLESEED_RENDERER_KERNEL_LIGHTING_BACKWARDLIGHTSAMPLER_H
+
+// appleseed.renderer headers.
+#include "renderer/kernel/intersection/intersectionsettings.h"
+#include "renderer/kernel/lighting/lightsample.h"
+#include "renderer/kernel/lighting/lighttree.h"
+#include "renderer/kernel/lighting/lighttypes.h"
+#include "renderer/kernel/shading/shadingray.h"
+#include "renderer/modeling/scene/containers.h"
+#include "renderer/utility/transformsequence.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/concepts/noncopyable.h"
+#include "foundation/math/cdf.h"
+#include "foundation/math/hash.h"
+#include "foundation/math/transform.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/types.h"
+#include "foundation/utility/containers/hashtable.h"
+#include "foundation/utility/uid.h"
+
+// Standard headers.
+#include <cstddef>
+#include <vector>
+
+// Forward declarations.
+namespace renderer  { class Assembly; }
+namespace renderer  { class AssemblyInstance; }
+namespace renderer  { class Intersector; }
+namespace renderer  { class Light; }
+namespace renderer  { class Material; }
+namespace renderer  { class MaterialArray; }
+namespace renderer  { class ObjectInstance; }
+namespace renderer  { class ParamArray; }
+namespace renderer  { class Scene; }
+namespace renderer  { class ShadingPoint; }
+
+namespace renderer
+{
+
+//
+// The light sampler collects all the light-emitting entities (non-physical lights, mesh lights)
```

Comment:
We need to update this comment to also explain what this light sampler is used for.